tcl8.3.4/doc/Thread.3
'\"
'\" Copyright (c) 1999 Scriptics Corporation
'\" Copyright (c) 1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: Thread.3,v 1.1 2001/11/05 20:06:32 jgdavidson Exp $
'\" 
.so man.macros
.TH Threads 3 "8.1" Tcl "Tcl Library Procedures"
.BS
.SH NAME
Tcl_ConditionNotify, Tcl_ConditionWait, Tcl_ConditionFinalize, Tcl_GetThreadData, Tcl_MutexLock, Tcl_MutexUnlock, Tcl_MutexFinalize, Tcl_CreateThread \- Tcl thread support.
.SH SYNOPSIS
.nf
\fB#include <tcl.h>\fR
.sp
void
\fBTcl_ConditionNotify\fR(\fIcondPtr\fR)
.sp
void
\fBTcl_ConditionWait\fR(\fIcondPtr, mutexPtr, timePtr\fR)
.sp
void
\fBTcl_ConditionFinalize\fR(\fIcondPtr\fR)
.sp
Void *
\fBTcl_GetThreadData\fR(\fIkeyPtr, size\fR)
.sp
void
\fBTcl_MutexLock\fR(\fImutexPtr\fR)
.sp
void
\fBTcl_MutexUnlock\fR(\fImutexPtr\fR)
.sp
void
\fBTcl_MutexFinalize\fR(\fImutexPtr\fR)
.sp
int
\fBTcl_CreateThread\fR(\fIidPtr, threadProc, clientData, stackSize, flags\fR)
.SH ARGUMENTS
.AS Tcl_ThreadDataKey *keyPtr
.AP Tcl_Condition *condPtr in
A condition variable, which must be associated with a mutex lock.
.AP Tcl_Condition *mutexPtr in
A mutex lock.
.AP Tcl_Time *timePtr in
A time limit on the condition wait.  NULL to wait forever.
Note that a polling value of 0 seconds doesn't make much sense.
.AP Tcl_ThreadDataKey *keyPtr in
This identifies a block of thread local storage.  The key should be
static and process-wide, yet each thread will end up associating
a different block of storage with this key.
.AP int *size in
The size of the thread local storage block.  This amount of data
is allocated and initialized to zero the first time each thread
calls \fBTcl_GetThreadData\fR.
.AP Tcl_ThreadId *idPtr out
The refered storage will contain the id of the newly created thread as
returned by the operating system.
.AP Tcl_ThreadId id in
Id of the thread waited upon.
.AP Tcl_ThreadCreateProc threadProc in
This procedure will act as the \fBmain()\fR of the newly created
thread. The specified \fIclientData\fR will be its sole argument.
.AP ClientData clientData in
Arbitrary information. Passed as sole argument to the \fIthreadProc\fR.
.AP int stackSize in
The size of the stack given to the new thread.
.AP int flags in
Bitmask containing flags allowing the caller to modify behaviour of
the new thread.
.AP int *result out
The refered storage is used to place the exit code of the thread
waited upon into it.
.BE
.SH INTRODUCTION
Beginning with the 8.1 release, the Tcl core is thread safe, which
allows you to incorporate Tcl into multithreaded applications without
customizing the Tcl core.  To enable Tcl multithreading support,
you must include the \fB--enable-threads\fR option to \fBconfigure\fR
when you configure and compile your Tcl core.
.PP
An important contstraint of the Tcl threads implementation is that
\fIonly the thread that created a Tcl interpreter can use that
interpreter\fR.  In other words, multiple threads can not access
the same Tcl interpreter.  (However, as was the case in previous
releases, a single thread can safely create and use multiple
interpreters.)
.PP
.VS 8.3.1
Tcl does provide \fBTcl_CreateThread\fR for creating threads. The
caller can determine the size of the stack given to the new thread and
modify the behaviour through the supplied \fIflags\fR. The value
\fBTCL_THREAD_STACK_DEFAULT\fR for the \fIstackSize\fR indicates that
the default size as specified by the operating system is to be used
for the new thread. As for the flags, currently are only the values
\fBTCL_THREAD_NOFLAGS\fR and \fBTCL_THREAD_JOINABLE\fR defined. The
first of them invokes the default behaviour with no
specialities. Using the second value marks the new thread as
\fIjoinable\fR. This means that another thread can wait for the such
marked thread to exit and join it.
.PP
Restrictions: On some unix systems the pthread-library does not
contain the functionality to specify the stacksize of a thread. The
specified value for the stacksize is ignored on these systems. Both
Windows and Macintosh currently do not support joinable threads. This
flag value is therefore ignored on these platforms.
.VE
.PP
Tcl does provide \fBTcl_ExitThread\fR and \fBTcl_FinalizeThread\fR
for terminating threads and invoking optional per-thread exit
handlers.  See the \fBTcl_Exit\fR page for more information on these
procedures.
.PP
Tcl provides \fBTcl_ThreadQueueEvent\fR and \fBTcl_ThreadAlert\fR
for handling event queueing in multithreaded applications.  See
the \fBNotifier\fR manual page for more information on these procedures.
.PP
In this release, the Tcl language itself provides no support for
creating multithreaded scripts (for example, scripts that could spawn
a Tcl interpreter in a separate thread).  If you need to add this
feature at this time, see the \fItclThreadTest.c\fR
file in the Tcl source distribution for an experimental implementation
of a Tcl "Thread" package implementing thread creation and management
commands at the script level.

.SH DESCRIPTION
A mutex is a lock that is used to serialize all threads through a piece
of code by calling \fBTcl_MutexLock\fR and \fBTcl_MutexUnlock\fR.
If one thread holds a mutex, any other thread calling \fBTcl_MutexLock\fR will
block until \fBTcl_MutexUnlock\fR is called.
.VS
A mutex can be destroyed after its use by calling \fBTcl_MutexFinalize\fR.
The result of locking a mutex twice from the same thread is undefined.
On some platforms it will result in a deadlock.
.VE
The \fBTcl_MutexLock\fR, \fBTcl_MutexUnlock\fR and \fBTcl_MutexFinalize\fR
procedures are defined as empty macros if not compiling with threads enabled.
.PP
A condition variable is used as a signaling mechanism:
a thread can lock a mutex and then wait on a condition variable
with \fBTcl_ConditionWait\fR.  This atomically releases the mutex lock
and blocks the waiting thread until another thread calls
\fBTcl_ConditionNotify\fR.  The caller of \fBTcl_ConditionNotify\fR should
have the associated mutex held by previously calling \fBTcl_MutexLock\fR,
but this is not enforced.  Notifying the
condition variable unblocks all threads waiting on the condition variable,
but they do not proceed until the mutex is released with \fBTcl_MutexUnlock\fR.
The implementation of \fBTcl_ConditionWait\fR automatically locks
the mutex before returning.
.PP
The caller of \fBTcl_ConditionWait\fR should be prepared for spurious
notifications by calling \fBTcl_ConditionWait\fR within a while loop
that tests some invariant.
.PP
.VS
A condition variable can be destroyed after its use by calling
\fBTcl_ConditionFinalize\fR.
.PP
The \fBTcl_ConditionNotify\fR, \fBTcl_ConditionWait\fR and
\fBTcl_ConditionFinalize\fR procedures are defined as empty macros if
not compiling with threads enabled.
.VE
.PP
The \fBTcl_GetThreadData\fR call returns a pointer to a block of
thread-private data.  Its argument is a key that is shared by all threads
and a size for the block of storage.  The storage is automatically 
allocated and initialized to all zeros the first time each thread asks for it.
The storage is automatically deallocated by \fBTcl_FinalizeThread\fR.
.SH INITIALIZATION
.PP
All of these synchronization objects are self initializing.
They are implemented as opaque pointers that should be NULL
upon first use.
The mutexes and condition variables are
.VS
either cleaned up by process exit handlers (if living that long) or
explicitly by calls to \fBTcl_MutexFinalize\fR or
\fBTcl_ConditionFinalize\fR.
.VE
Thread local storage is reclaimed during \fBTcl_FinalizeThread\fR.
.SH "CREATING THREADS"
The API to create threads is not finalized at this time.
There are private facilities to create threads that contain a new
Tcl interpreter, and to send scripts among threads.
Dive into tclThreadTest.c and tclThread.c for examples.
.SH "SEE ALSO"
Tcl_GetCurrentThread, Tcl_ThreadQueueEvent, Tcl_ThreadAlert,
Tcl_ExitThread, Tcl_FinalizeThread,
Tcl_CreateThreadExitHandler, Tcl_DeleteThreadExitHandler
.SH KEYWORDS
thread, mutex, condition variable, thread local storage


ChangeLog
2004-06-14 20:28  Dossy Shiobara <dossy@panoptic.com>

	* nsd/connio.c: If nsend == 0, we would never call Ns_WriteConn to
	  flush the queued headers.  This could happen when we're sending
	  zero bytes of data as a response.  This fixes bug #971016 in the
	  case where fastpath.cache=false and fastpath.mmap=false and a
	  zero byte file is requested.

2004-06-14 20:25  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: unnecessary test for data != NULL actually caused
	  part of bug #971016 when fastpath.cache=false and
	  fastpath.mmap=true and the file requested is zero bytes, mmap()
	  returns 0 which gets passed along as data == NULL, causing
	  ReturnCharData() to not flush the queued headers.  removing the
	  if is safe as Ns_WriteConn will simply flush any queued data

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* Makefile: remove Makefile.module on distclean.
	* include/ns.h, nsd/log.c: Allow users to override logging functions.
	* nsd/str.c: Fix crash bug in Ns_Trim* when trimming NULL strings.

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* TAG aolserver_v40_r5

2004-03-09 13:35  Rob Crittenden <rcrittenden0569@aol.com>

	* nsd/tclhttp.c: The last header was being missed because the pointer
	  was being advanced past the \r\n.

2004-02-26 14:37  Dossy Shiobara <dossy@panoptic.com>

	* nsd/unix.c: uid_t is supposed to be unsigned, but apparently
	  isn't on OSX, apparently.

2004-02-25 19:56  Dossy Shiobara <dossy@panoptic.com>

	* nscgi/nscgi.c: Ns_ConnSetRequiredHeaders now requires length < 0
	  if you don't want a Content-Length header

2004-02-25 19:55  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: fix so that it's still possible to formulate a
	  response with no Content-Length header (such as for nscgi, or
	  HTTP streaming) where you don't know the Content-Length header in
	  advance

2004-02-24 07:15  Dossy Shiobara <dossy@panoptic.com>

	* nsd/: nsmain.c, nsd.h, driver.c: fix race condition where
	  DriverThreads are spawned but before they can Ns_SockListenEx,
	  the main thread does its NsClosePreBound and closes the prebound
	  sockets, making them unavailable to the DriverThreads if they
	  require root priv to listen on

2004-02-18 01:30  Dossy Shiobara <dossy@panoptic.com>

	* nsperm/nsperm.c: fixed bug #899364, restoring old nsperm
	  functionality as it was in 3.x.

2004-02-15  Jim Davidson <jgdavidson@aol.com>

	* nsd/conn.c:
	* include/ns.h:
	* nsd/return.c: Added new Ns_ConnGet/SetKeepAliveFlag APIs to
	allow custom request procedures to enable keep-alive directly.

	* nsd/driver.c: Minor updates to separate conn & driver structs.

2004-01-01  Scott S. Goodwin <scott@scottg.net>

        * nsd/adpparse.c: fixed bug #868600, where tag attribute values
	encapsulated within single quotes weren't being parsed properly.

2003-12-27  Scott S. Goodwin <scott@scottg.net>

        * nsd/urlopen.c: fixed bug #844809.

        * sample-config.tcl: fixed bug #805761.

2003-12-24  Scott S. Goodwin <scott@scottg.net>

	* include/nsthread.h: Bumped up NS_THREAD_NAMESIZE from 32 to 64.
	Mutex names were being truncated, making it impossible to distinguish
	mutexes with longer names when looking at stats.

        * nscgi/nscgi.c: Changed HTTPS env variable to be set to "on" or "off"
	instead of "1" or "0". There doesn't appear to be any standard for
	this, but Apache uses the on/off string in the environment.

2003-12-16  Scott S. Goodwin <scott@scottg.net>

        * nscgi/nscgi.c: Fixed issue with HTTPS environment variable not being
        set, and port not being set to 443 if connection is SSL'd. Manually
        applied these changes from a patch supplied by Vinod Kurup.

2003-11-16  Jim Davidson <jgdavidson@aol.com>

	* include/ns.h: Update version ids, added NS_SOCK_DROP and
	NS_TCL_TRACE constants and pre-queue function definitions.

	* nsd/nsd.h:  Added new "times" struct to Conn for more
	fine-grained timing.  Need to provide options to enable
	these time stamps plus ways to get the data.

	* nsd/driver.c:  Restructured in a few ways:

	1. Each loaded driver now has it's own thread.  This should
	allow some more concurrency (if necessary) and generally
	made the code more manageable.

	2. New pre-queue I/O callback API.  The API will be documented
	with an example HTTP module later.

	3. Conn allocation is handled by the driver threads and
	request buffers have been moved to the Conn from the Sock.

	4. Reads for synchrounous drivers (i.e., SSL) are handled
	by a pool of reader threads instead of by the conn threads.
	The number of threads is set with a "maxreaders" config
	option.  This change should enable both Host: header based
	virtual servers and pre-queue callbacks for SSL connections.

	5. New "maxsocks" config for drivers is now used to manage
	the concurrency limit of the server.  Conn's are now queued
	without limit into the connection pools and the previous
	"maxconns" virtual-server config is ignored.

	* nsd/queue.c: Cleaned up "connsperthread" code and updated
	NsQueueConn to support new driver code.

	* nsd/conn.c:
	* nsd/connio.c:
	* nsd/form.c:
	* nsd/nsd.h:
	* nsd/queue.c:
	* nsd/return.c: Moved connection request buffers from Sock
	to Conn structure.

	* nsd/server.c: Removed startup Conn pre-allocation as Conns
	are now allocated dynamically by the driver.  Also, removed
	the "maxconns" parameter which is no longer used.

	* nssock/nssock.c: Added new "async" option (default on)
	mostly to test new reader thread capabilities of the core
	driver code.

	* nsd/tclinit.c: Added a new generalized Tcl interp tracing
	facility and cleaned up some previous init code.  This
	should make it easier to support standard Tcl extensions
	and will be documented with an example module later.

	* nsd/nsmain.c: Fixed order of driver shutdown wait.

	* nsd/init.c: Fixed order of init calls.

	* nsd/binder.c:
	* nsd/dns.c: Named unnamed mutexes.

	* nsd/tclset.c:
	* nsd/tclsock.c: Removed unused functions.


nsd/thread.c
/*
 * The contents of this file are subject to the AOLserver Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://aolserver.com/.
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
 * the License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is AOLserver Code and related documentation
 * distributed by AOL.
 * 
 * The Initial Developer of the Original Code is America Online,
 * Inc. Portions created by AOL are Copyright (C) 1999 America Online,
 * Inc. All Rights Reserved.
 *
 * Alternatively, the contents of this file may be used under the terms
 * of the GNU General Public License (the "GPL"), in which case the
 * provisions of GPL are applicable instead of those above.  If you wish
 * to allow use of your version of this file only under the terms of the
 * GPL and not to allow others to use your version of this file under the
 * License, indicate your decision by deleting the provisions above and
 * replace them with the notice and other provisions required by the GPL.
 * If you do not delete the provisions above, a recipient may use your
 * version of this file under either the License or the GPL.
 */


/* 
 * thread.c --
 *
 *	Routines for creating, exiting, and joining threads.
 */

static const char *RCSID = "@(#) $Header: /Users/dossy/Desktop/cvs/aolserver/nsd/Attic/thread.c,v 1.1 2001/11/05 20:26:51 jgdavidson Exp $, compiled: " __DATE__ " " __TIME__;

#include "nsd.h"
#include <sched.h>		/* sched_yield() */

/*
 * The following constants define the default and minimum stack
 * sizes for new threads.
 */

#define STACK_DEFAULT	65536	/* 64k */
#define STACK_MIN	16384	/* 16k */

/*
 * The following defines the estimated stack space required to
 * return an NS_OK in Ns_CheckStack().
 */

#define STACK_CHECK	2048	/* 2k */

/*
 * The following structure maintains all state for a thread
 * including thread local storage slots.
 */

typedef struct Thread {
    struct Thread  *nextPtr;	/* Next in list of all threads. */
    time_t	    ctime;	/* Thread structure create time. */
    int		    flags;	/* Detached, joined, etc. */
    Ns_ThreadProc  *proc;	/* Thread startup routine. */ 
    void           *arg;	/* Argument to startup proc. */
    int		    tid;        /* Id set by thread for logging. */
    long	    stackSize;	/* Stack size in bytes for this thread. */
    void	   *stackBase;	/* Approximate stack base for Ns_CheckStack. */
    char	    name[NS_THREAD_NAMESIZE+1]; /* Thread name. */
    char	    parent[NS_THREAD_NAMESIZE+1]; /* Parent name. */
}               Thread;

static void *ThreadMain(void *arg);
static Thread *NewThread(void);
static Thread *GetThread(void);
static void SetThread(Thread *thrPtr);
static void CleanupThread(void *arg);

/*
 * The following pointer maintains a linked list of all threads.
 */

static Thread *firstThreadPtr;

/*
 * The following maintains the tls key for the thread context.
 */

static Ns_Tls key;


/*
 *----------------------------------------------------------------------
 *
 * NsInitThreads --
 *
 *	Initialize threads interface.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Allocates pthread_key_t for thread context.
 *
 *----------------------------------------------------------------------
 */

void
NsInitThreads(void)
{
    Ns_TlsAlloc(&key, CleanupThread);
    NsInitMaster();
    NsInitReentrant();
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadCreate --
 *
 *	Create a new thread thread.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	A new thread is allocated and started.
 *
 *----------------------------------------------------------------------
 */

void
Ns_ThreadCreate(Ns_ThreadProc *proc, void *arg, long stack,
    	    	Ns_Thread *resultPtr)
{
    static char *func = "Ns_ThreadCreate";
    pthread_attr_t attr, *attrPtr;
    pthread_t pid;
    Thread *thrPtr;
    int     err;

    /*
     * Determine the stack size and impose a 16k minimum.
     */

    if (stack == 0) {
	stack = nsconf.stacksize;
    }
    if (stack < STACK_MIN) {
	stack = STACK_MIN;
    }

    /*
     * Allocate a new thread structure and update values
     * which are known for threads created here.
     */

    Ns_MasterLock();
    thrPtr = NewThread();
    thrPtr->proc = proc;
    thrPtr->arg = arg;
    thrPtr->stackSize = stack;
    if (resultPtr == NULL) {
    	thrPtr->flags = NS_THREAD_DETACHED;
    }
    strcpy(thrPtr->parent, Ns_ThreadGetName());
    Ns_MasterUnlock();

    /*
     * Set the stack size.  It could be smarter to leave the default 
     * on platforms which map large stacks with guard zones
     * (e.g., Solaris and Linux).
     */

    err = pthread_attr_init(&attr);
    if (err != 0) {
        NsThreadFatal(func, "pthread_attr_init", err);
    }
    err = pthread_attr_setstacksize(&attr, thrPtr->stackSize); 
    if (err != 0) {
        NsThreadFatal(func, "pthread_attr_setstacksize", err);
    }

    /*
     * System scope threads are used by default on Solaris and UnixWare.
     * Other platforms are either already system scope (Linux, HP11), 
     * support only process scope (SGI), user-level thread libraries
     * anyway (FreeBSD), or found to be unstable with this setting (OSF).
     */

#ifdef USE_PTHREAD_SYSSCOPE
    err = pthread_attr_setscope(&attr, PTHREAD_SCOPE_SYSTEM);
    if (err != 0) {
        NsThreadFatal(func, "pthread_attr_setscope",err);
    }
#endif
    err = pthread_create(&pid, &attr, ThreadMain, thrPtr);
    if (err != 0) {
        NsThreadFatal(func, "pthread_create", err);
    }
    err = pthread_attr_destroy(&attr);
    if (err != 0) {
        NsThreadFatal(func, "pthread_attr_destroy", err);
    }
    if (resultPtr != NULL) {
	*resultPtr = (Ns_Thread) pid;
    } else {
    	err = pthread_detach(pid);
        if (err != 0) {
            NsThreadFatal(func, "pthread_detach", err);
        }
    }
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadExit --
 *
 *	Exit the thread.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Exit arg will be passed through pthread_exit.
 *
 *----------------------------------------------------------------------
 */

void
Ns_ThreadExit(void *arg)
{
    pthread_exit(arg);
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadJoin --
 *
 *	Wait for exit of a non-detached thread.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Requested thread is destroyed after join.
 *
 *----------------------------------------------------------------------
 */

void
Ns_ThreadJoin(Ns_Thread *threadPtr, void **argPtr)
{
    pthread_t pid = (pthread_t) *threadPtr;
    int err;

    err = pthread_join(pid, argPtr);
    if (err != 0) {
	NsThreadFatal("Ns_ThreadJoin", "pthread_join", err);
    }
}


/*
 *----------------------------------------------------------------------
 *
 * ThreadMain --
 *
 *	Thread startup routine.  Sets the given pre-allocated thread
 *	structure and calls the user specified procedure.
 *
 * Results:
 *	None.  Will call Ns_ThreadExit if not called by the
 *	user code.
 *
 * Side effects:
 *	See SetThread.
 *
 *----------------------------------------------------------------------
 */

static void *
ThreadMain(void *arg)
{
    Thread      *thrPtr = (Thread *) arg;
    char	 name[NS_THREAD_NAMESIZE];

    SetThread(thrPtr);
    sprintf(name, "-thread%d-", thrPtr->tid);
    Ns_ThreadSetName(name);
    thrPtr->stackBase = &arg;
    (*thrPtr->proc) (thrPtr->arg);
    return NULL;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadId --
 *
 *	Return the numeric thread id for calling thread.
 *
 * Results:
 *	Integer thread id.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

int
Ns_ThreadId(void)
{
    return (int) pthread_self();
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadSelf --
 *
 *	Return opaque handle to thread's data structure.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Value at thrPtr is updated with thread's data structure pointer.
 *
 *----------------------------------------------------------------------
 */

void
Ns_ThreadSelf(Ns_Thread *threadPtr)
{
    *threadPtr = (Ns_Thread) pthread_self();
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_CheckStack --
 *
 *	Check for possible thread stack overflow.
 *
 * Results:
 *	NS_OK if stack appears ok, otherwise NS_ERROR.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

int
Ns_CheckStack(void)
{
    Thread *thisPtr = GetThread();
    long size, base, here;
    
    /* 
     * Check to see if the thread may be about to grow beyond it's allocated
     * stack.  Currently, this function is only called in Tcl_Eval where
     * it traps the common case of infinite Tcl recursion.  The STACK_CHECK
     * slop is just a conservative guess.
     */

    size = thisPtr->stackSize - STACK_CHECK;
    base = (long) thisPtr->stackBase;
    here = (long) &thisPtr;
    if (size > 0 && abs(base - here) > size) {
	return NS_ERROR;
    }
    return NS_OK;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadGetName --
 *
 *	Return a pointer to calling thread's string name.
 *
 * Results:
 *	Pointer to thread name string.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

char *
Ns_ThreadGetName(void)
{
    Thread *thisPtr = GetThread();

    return thisPtr->name;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadSetName --
 *
 *	Set the name of the calling thread.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	String is copied to thread data structure.
 *
 *----------------------------------------------------------------------
 */

void
Ns_ThreadSetName(char *name)
{
    Thread *thisPtr = GetThread();

    Ns_MasterLock();
    strncpy(thisPtr->name, name, NS_THREAD_NAMESIZE);
    Ns_MasterUnlock();
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadGetParent --
 *
 *	Return a pointer to calling thread's parent name.
 *
 * Results:
 *	Pointer to thread parent name string.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

char *
Ns_ThreadGetParent(void)
{
    Thread *thisPtr = GetThread();

    return thisPtr->parent;
}


/*
 *----------------------------------------------------------------------
 *
 * NsThreadInfo --
 *
 *	Append info for each thread.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *  	None.
 *
 *----------------------------------------------------------------------
 */

void
NsThreadInfo(Tcl_DString *dsPtr)
{
    Thread *thrPtr;
    char buf[100];

    Ns_MasterLock();
    thrPtr = firstThreadPtr;
    while (thrPtr != NULL) {
	Tcl_DStringStartSublist(dsPtr);
	Tcl_DStringAppendElement(dsPtr, thrPtr->name);
	Tcl_DStringAppendElement(dsPtr, thrPtr->parent);
	sprintf(buf, " %d %d %ld", thrPtr->tid, thrPtr->flags, thrPtr->ctime);
	Tcl_DStringAppend(dsPtr, buf, -1);
	Ns_GetProcInfo(dsPtr, (void *) thrPtr->proc, thrPtr->arg);
	Tcl_DStringEndSublist(dsPtr);
	thrPtr = thrPtr->nextPtr;
    }
    Ns_MasterUnlock();
}


/*
 *----------------------------------------------------------------------
 *
 * NewThread --
 *
 *	Allocate a new thread data structure and add it to the list
 *	of all threads.  The new thread is suitable for a detached,
 *	unknown thread such as the initial thread but Ns_ThreadCreate
 *	will update as necessary before creating the new threads.
 *
 * Results:
 *	Pointer to new Thread.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

static Thread *
NewThread(void)
{
    Thread *thrPtr;

    thrPtr = ns_calloc(1, sizeof(Thread));
    thrPtr->ctime = time(NULL);
    Ns_MasterLock();
    thrPtr->nextPtr = firstThreadPtr;
    firstThreadPtr = thrPtr;
    Ns_MasterUnlock();
    return thrPtr;
}


/*
 *----------------------------------------------------------------------
 *
 * SetThread --
 *
 *	Pthread routine to set this thread's nsthread data structure
 *	and update the tid.
 *
 * Results:
 *	None
 *
 * Side effects:
 *	Key is allocated the first time.
 *
 *----------------------------------------------------------------------
 */

static void
SetThread(Thread *thrPtr)
{
    int err;

    thrPtr->tid = (int) pthread_self();
    Ns_TlsSet(&key, thrPtr);
}


/*
 *----------------------------------------------------------------------
 *
 * GetThread --
 *
 *	Return this thread's nsthread data structure, initializing
 *	it if necessary, normally for the first thread but also
 *	for threads created without Ns_ThreadCreate.
 *
 * Results:
 *	Pointer to per-thread data structure.
 *
 * Side effects:
 *	Key is allocated the first time.
 *
 *----------------------------------------------------------------------
 */

static Thread      *
GetThread(void)
{
    Thread *thisPtr;

    thisPtr = Ns_TlsGet(&key);
    if (thisPtr == NULL) {
	thisPtr = NewThread();
    	thisPtr->flags = NS_THREAD_DETACHED;
	SetThread(thisPtr);
    }
    return thisPtr;
}


/*
 *----------------------------------------------------------------------
 *
 * Ns_ThreadYield --
 *
 *	Yield the cpu to another thread.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	See sched_yield().
 *
 *----------------------------------------------------------------------
 */

void
Ns_ThreadYield(void)
{
    sched_yield();
}


/*
 *----------------------------------------------------------------------
 *
 * CleanupThread --
 *
 *	TLS cleanup for the nsthread context. 
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

static void
CleanupThread(void *arg)
{
    Thread **thrPtrPtr;
    Thread *thrPtr = arg;

    Ns_MasterLock();
    thrPtrPtr = &firstThreadPtr;
    while (*thrPtrPtr != thrPtr) {
	thrPtrPtr = &(*thrPtrPtr)->nextPtr;
    }
    *thrPtrPtr = thrPtr->nextPtr;
    thrPtr->nextPtr = NULL;
    Ns_MasterUnlock();
    ns_free(thrPtr);
}
